### PR TITLE
Set registration email default to false

### PR DESCRIPTION
**Pull request by dmint789**
Addresses issue https://github.com/thewca/worldcubeassociation.org/issues/4673. This simply sets the default to registration emails for delegates and organizers to false, so that they can opt in to receiving these emails if they so choose.
Please review that I implemented the change correctly.

### CODE REVIEW

**Comment by denthebro** [comment]
Oops, looks like I did something wrong...

**Comment by saranshgrover** [comment]
Please update the [test](https://github.com/thewca/worldcubeassociation.org/blob/master/WcaOnRails/spec/models/competition_spec.rb#L616) accordingly.

File: WcaOnRails/app/models/competition.rb
Comment (jonatanklosko) [inline WcaOnRails/app/models/competition.rb:702]: If you look at the `registrations_mailer.rb`:

https://github.com/thewca/worldcubeassociation.org/blob/e2880132491b1eeee5d4a1b04701d6f34f71e9c9/WcaOnRails/app/mailers/registrations_mailer.rb#L8-L13

you can see we use the `receive_registration_emails` attribute (database column) directly to determine email recipients.

The `receiving_registration_emails?` only determines the current setting for the given user.

If we want to change the default, it should be done on the database level via migration. Currently the default for `receive_registration_emails` is `true` (you can look it up in the `structure.sql` file).

Comment (denthebro) [reply WcaOnRails/app/models/competition.rb:702]: Thank you for the tip, this'll make it much easier

**Comment by denthebro** [comment]
I think I have fixed it. I just went with @jonatanklosko's suggestion. Haven't tested if it's working though, as I have never worked with catching emails.

**Comment by jonatanklosko** [comment]
Updating the file alone has no effect on the database, instead we need a migration. You can see a bunch of past migrations [here](https://github.com/thewca/worldcubeassociation.org/tree/master/WcaOnRails/db/migrate), you can think of each migration as a script that modifies the actual database. This approach allows for applying changes to the production database, as well as the local database of every developer simply by running new migrations.

You can generate a new migration file with `bin/rails g migration change_receive_registration_emails_default` and usually define a `change` method (sometimes you may want separate `up` and `down` methods if you do less trivial things). For pretty much any kind of database structure change there's a [built-in method](https://apidock.com/rails/v5.2.3/ActiveRecord/ConnectionAdapters/SchemaStatements/change_column_default).

Once you write the migration, you can run it locally with `bin/rails db:migrate`. If successful, this should apply the change to your local database and also automatically update `structure.sql` accordingly. There's also `bin/rails db:rollback` if you want to revert the migration.

When running in Docker, you need to `exec` these commands in the container, similarly to how you'd run tests and other stuff.

**Comment by gregorbg** [comment]
Do you need any more guidance in writing the migration for this PR?

**Comment by denthebro** [comment]
> Do you need any more guidance in writing the migration for this PR?

I'm not working on this PR at the moment. I'll try to come back to this when my schedule frees up a bit.

**Comment by dmint789** [comment]
@gregorbg a lot of time has passed since I opened this PR, so I was wondering if this issue has been dealt with elsewhere by now. If not, I'll make the change asap.

**Comment by dmint789** [comment]
Please remind me how to run the migration on production. Also am I only meant to do it after this PR is merged?

File: WcaOnRails/db/migrate/20220529123854_email_default_change.rb
Comment (gregorbg) [inline WcaOnRails/db/migrate/20220529123854_email_default_change.rb:1]: Please give a more concise name to the migration. Which email default is being changed?

Comment (gregorbg) [inline WcaOnRails/db/migrate/20220529123854_email_default_change.rb:1]: Also, see other migrations for a `frozen_string_literal` comment that has to be at the top of the file, otherwise RuboCop starts crying like a baby...

**Review by gregorbg — CHANGES_REQUESTED** [review]
The technicalities of your migration content are fine. However, you have to execute the migration locally which will change/affect the `structure.sql` file. The changes to that file have to be checked in as part of this PR too, otherwise the CI fails.

Migrations on prod are run manually by me or other Senior Members when deployments are done.

We should probably also announce this in an email to delegates@ because this is a substantial change in competititon workflow. Please prepare a draft in our Google Drive and reference it in a comment here before merging :)

File: WcaOnRails/db/migrate/20220529123854_email_default_change.rb
Comment (dmint789) [reply WcaOnRails/db/migrate/20220529123854_email_default_change.rb:1]: Do you mean a more descriptive name? Making it more concise would make it hard to specify which default is being changed.

**Comment by dmint789** [comment]
I actually _have_ run the migration locally, but I intentionally avoided committing the changes to structure.sql, as you told me not to do this last time I made a PR with a migration. Are you sure you want me to do it this time?

**Comment by gregorbg** [comment]
For the record, Deni's question has been resolved in direct conversation. Gist: Only commit the changes to `structure.sql` that are actually relevant

**Comment by dmint789** [comment]
@gregorbg 
> We should probably also announce this in an email to delegates@ because this is a substantial change in competititon workflow. Please prepare a draft in our Google Drive and reference it in a comment here before merging :)

I have created the email draft in our Google Drive. Should I copy the text or paste a link to it?

**Comment by dmint789** [comment]
Oh no, another migration has been created since I started this PR. I will redo my commit after a rebase

**Comment by dmint789** [comment]
I think I have now fixed the migration

**Comment by gregorbg** [comment]
Please verify that further up in the `structure.sql` file, all of the actual column defaults have been changed in the individual tables' SQL `CREATE TABLE` definitions as well

**Comment by dmint789** [comment]
@gregorbg just to make sure I understand correctly, the column defaults don't make a difference since the migration changes the value anyways, right? We're just doing this for the sake of elegance?

**Comment by gregorbg** [comment]
Essentially, yes. The `test` environment that is used to run all unit and integration tests in our CI also uses the `structure.sql` file to create a mock database from scratch. So keeping that file up to date helps with avoiding headaches about inconsistent behavior while testing as well :)

**Comment by dmint789** [comment]
Ugh, now the test has to be updated too, cause it's not expecting that value to be 0

**Comment by gregorbg** [comment]
Deleting tests is never a good idea as long as the basic feature they're supposed to test still exists.

In this case, the mail sending feature still exists, we just need to take a different route to trigger them. I suggest two things:
1. Adjusting the existing tests so that they create the competition with the flag set to `true`, and they can then expect the mail to be sent.
2. Create a second set of tests by copy/pasting (relevant parts of) the original tests and change them to expect that an email is NOT being sent. Covers both cases and gives us more confidence in our test suite.

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [inline WcaOnRails/app/models/competition.rb:750]: Why are you negating the condition here?
The purpose of this method is to check whether a user should receive registration emails. Your updated check is basically saying "If the flag `receive_registration_emails` is NOT set, reply that they should receive emails" which sounds very wrong 🙈

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [inline WcaOnRails/app/models/competition.rb:754]: Negated condition, see other comment above.

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [inline WcaOnRails/app/models/competition.rb:758]: Negated condition, see other comment above.

**Comment by gregorbg** [comment]
Uh oh, wrong keybinding. I am very sorry for accidentally closing this.

**Review by gregorbg — CHANGES_REQUESTED** [review]

**Comment by dmint789** [comment]
I have looked into the tests and it looks like we actually already have tests that check that no email is sent when the checkbox is set to false, so I don't believe any new tests will be necessary. I will update the existing tests shortly.

**Review by gregorbg — APPROVED** [review]
Good job on figuring out the tests! LGTM but please don't merge until we can coordinate the deployment + email sendout (reach out on Slack)

**Pull request merged**
